nethack/src/board.c

```diff
@@ -0,0 +1,253 @@
+/* vim:set cin ft=c sw=4 sts=4 ts=8 et ai cino=Ls\:0t0(0 : -*- mode:c;fill-column:80;tab-width:8;c-basic-offset:4;indent-tabs-mode:nil;c-file-style:"k&r" -*-*/
+/* NetHack may be freely redistributed.  See license for details. */
+
+#include "nhcurses.h"
+#include <time.h>
+
+static struct gamewin *gw;
+static struct win_menu *mdat;
+static nh_bool done;
+static struct nh_board_entry *entries;
+static int listlen, maxplayersize;
+
+static void
+board_display_curses_ui(struct nh_menuitem *items, int icount)
+{
+    int startx, starty, x1, x2, y1, y2;
+
+    if (gw != NULL) {
+        delwin(mdat->content);
+        delwin(gw->win);
+        delete_gamewin(gw);
+    }
+
+    gw = alloc_gamewin(sizeof (struct win_menu));
+    gw->draw = draw_menu;
+    gw->resize = resize_menu;
+    mdat = (struct win_menu *)gw->extra;
+    mdat->items = items;
+    mdat->icount = icount;
+    mdat->title = "Binge Board";
+    mdat->selected = calloc(icount, sizeof (nh_bool));
+    if (!mdat->selected)
+        curses_msgwin("Couldn't allocate memory for the binge board");
+        return;
+    mdat->x1 = 0;
+    mdat->y1 = 0;
+    mdat->x2 = -1;
+    mdat->y2 = -1;
+
+    assign_menu_accelerators(mdat);
+    layout_menu(gw);
+
+    startx = (COLS - mdat->width) / 2;
+    starty = (LINES - mdat->height) / 2;
+
+    gw->win = newwin(mdat->height, mdat->width, starty, startx);
+    if (gw->win == NULL) {
```

Comment:
If this fails, on the next call to board_display_curses_ui it will call delwin(NULL) and crash.

There's a design pattern in C where you use `goto` to deallocate resources that you've allocated in case of an error. E.x.

```c
foo *complex_alloc() {
  foo *ptr = alloc_ptr();
  if (ptr == NULL) goto alloc_ptr_fail;
  ptr->bar = alloc_bar();
  if (ptr->bar == NULL) goto alloc_bar_fail;
  ptr->baz = alloc_baz();
  if (ptr->baz == NULL) goto alloc_baz_fail;

  return ptr;

alloc_baz_fail:
  free_bar(ptr->bar);
alloc_bar_fail:
  free_foo(ptr);
alloc_ptr_fail:
  return NULL;
}
```